language/puppet-functions.md

```diff
@@ -51,29 +51,60 @@ possible to call a puppet function with a lambda, and possibly also type a param
+
+#### Autoloading
+
+Puppet Functions are autoloaded from modules and the environment when located under <module-root>/functions or <environment-root>/functions respectively.
+
+The filename must match the simple (non namespaced name part) of the function. Thus, a function 'min' in module 'testmodule' is located like this:
 
     testmodule
       |- functions
          | min.pp
-         
-The functions defined this way in modules must be name-spaced. Thus the contents of this `min.pp` is:
+
+With the following contents in `min.pp` (note use of full namespace):
 
     function testmodule::min($a, $b) {
       # ...
     }
 
-It is allowed to defined Puppet Functions in the environment's logic (i.e. `site.pp`, or in the manifests loaded from the manifests directory). It is also allowed to defined functions directly
-on the command line when running puppet apply.
+And in an environment like this:
+
+    production
+      |- functions
+         | min.pp
+         
+With the following contents in `min.pp` (note use of namespace 'environment'):
+
+    function environment::min($a, $b) {
+      # ...
+    }
 
 Rules:
 
-* autoloaded functions in modules must have qualified names
-* functions that are defined in modules, but that are not autoloaded (loaded as part of something   
-  else) should also be defined in the module's namespace (it should be an error if an attempt is 
-  made to name it differently).
-* functions defined in the environment may have any name (global or namespaced)
-* A function loaded in the environment shadows functions from modules, but not system functions
+* Files containing an auto loaded function may only contain a single function.
+* Auto loading a function from a module does not trigger loading of the module's `init.pp` (nor is such initialization required to call a function from a module). If an author of a module provides functions that require that the module's init.pp is loaded, the function should include the module's class, or require that the caller first includes the module's class).
+
+##### Defining functions in Manifests
+
+It is possible to define a Puppet Functions in any manifest. Such functions will come into existence when the manifest in question is loaded for some reason other than calling the function (e.g. from 'site.pp' or when including a class).
+
+The following restrictions/rules/conventions apply on naming non-auto-loaded functions:
+
+* A function defined in a module **must** be qualified with the module's namespace
+* A function in an environment **should** begin with the special namespace 'environment'
+  * except when patching a function is required - then the name may shadow other functions defined in the same environment, or the modules in this environment, provided they have not already been auto loaded. Functions provided by the puppet runtime cannot be shadowed.
```

Comment:
Because code that is part of initial import is loaded first; essentially all code "in the environment" (site.pp) is loaded first. If you follow the rules (that are currently not enforced) you will not run into trouble. I guess that what happens if you break the rules is not that interesting as you are not supposed to break them.